### ISSUE TITLE
Spatial: Translate aggregate functions

### ISSUE THREAD

**Issue by bricelam** [dotnet/efcore#13278]
There are a handful of aggregate functions for spatial data. (Collect, Extent, Union, etc.)

It would be nice if we had a way to represent them in LINQ.

.NET                            | SQL Server                          | SpatiaLite                             | PostGIS
------------------------------- | ----------------------------------- | -------------------------------------- | -------
~~EnvelopeCombiner.Combine(geoms)~~ | EnvelopeAggregate(@​geoms)   | Extent(@​geoms)                 | ST_Extent(@​geoms)
GeometryCombiner.Combine(geoms) | CollectionAggregate(@​geoms) | Collect(@​geoms)                | ST_Collect(@​geoms)
UnaryUnionOp.Union(geoms)       | UnionAggregate(@​geoms)      | GUnion(@​geoms)                 | ST_Union(@​geoms)
ConvexHull.Create(geoms)        | ConvexHullAggregate(@​geoms) | ST_ConvexHull(Collect(@​geoms)) | ST_ConvexHull(ST_Collect(@​geoms))

**Comment by bricelam** [comment]
See also NetTopologySuite/NetTopologySuite#246

**Comment by roji** [comment]
As this is being proposed for 6.0, would we build more generic support for arbitrary aggregate functions? There's #11850 but that seems to be about DbFunctions - do we need a separate issue to track provider aggregate functions?

/cc @smitpatel

**Comment by smitpatel** [comment]
This is being proposed for 6.0? 😕

**Comment by roji** [comment]
Yeah, #22951...

**Comment by roji** [comment]
(a separate issue for tracking arbitrary non-DbFunction aggregate functions support may be a good idea regardless of 6.0, unless we just want to use #11850 to track that too)

**Comment by smitpatel** [comment]
>  arbitrary non-DbFunction aggregate functions support

They are already supported. 🤷

**Comment by roji** [comment]
Oh, nice!

**Comment by roji** [comment]
Specifically for plugins (NTS), note that we don't have an extension point for QueryableMethodTranslatingExpressionVisitor, which is needed if we want to support top-level invocations of aggregate functions (similar to Max/Sum/Average).

**Comment by bricelam** [comment]
@roji I added a table to the description in case you're feeling ambitious 😉

**Comment by roji** [comment]
Ahahahaha I'm definitely in a very aggregate mood

**Comment by roji** [comment]
Submitted PR #28115

Some notes:

* CollectionAggregate:
  * For two points (0,0) and (1,1), NTS [GeometryCombiner.Combine](https://nettopologysuite.github.io/NetTopologySuite/api/NetTopologySuite.Geometries.Utilities.GeometryCombiner.html) returns `MULTIPOINT ((0 0), (1 1))` whereas SQL Server [CollectionAggregate](https://docs.microsoft.com/en-us/sql/t-sql/spatial-geometry/collectionaggregate-geometry-data-type?view=sql-server-ver16) returns `GEOMETRYCOLLECTION (POINT (0 0), POINT (1 1))`. This is probably OK since NTS MultiPoint extends GeometryCollection, and GeometryCombiner.Combine's return type is Geometry.
  * A slightly more faithful translation for CollectionAggregate may be the GeometryCollection [constructor](https://nettopologysuite.github.io/NetTopologySuite/api/NetTopologySuite.Geometries.GeometryCollection.html#constructors), but unfortunately that accepts an array, not an IEnumerable (so we'd need to recognize and ignore ToArray over that). Translating constructor calls is also less easy.
* EnvelopeAggregate:
  * Weirdly, NTS [Envelope.Combiner.Combine](https://nettopologysuite.github.io/NetTopologySuite/api/NetTopologySuite.Geometries.Utilities.EnvelopeCombiner.html#NetTopologySuite_Geometries_Utilities_EnvelopeCombiner_Combine_System_Collections_Generic_IEnumerable_NetTopologySuite_Geometries_Geometry__) returns an [Envelope](https://nettopologysuite.github.io/NetTopologySuite/api/NetTopologySuite.Geometries.Envelope.html), which is not a Geometry. Note that [Geometry.Envelope](https://nettopologysuite.github.io/NetTopologySuite/api/NetTopologySuite.Geometries.Geometry.html#NetTopologySuite_Geometries_Geometry_Envelope) does return Geometry, which is why we can translate the non-aggregate version ([test](https://github.com/dotnet/efcore/blob/main/test/EFCore.Specification.Tests/Query/SpatialQueryTestBase.cs#L548)).
  * It's possible to convert an Envelope into a Geometry via [GeometryFactory.ToGeometry](https://nettopologysuite.github.io/NetTopologySuite/api/NetTopologySuite.Geometries.GeometryFactory.html#NetTopologySuite_Geometries_GeometryFactory_ToGeometry_NetTopologySuite_Geometries_Envelope_), but that makes the C#-side construct really complicated. We should probably wait for requests before looking into this.
* Aggregate convex hull:
  * SQL Server also has [ConvexHullAggregate](https://docs.microsoft.com/en-us/sql/t-sql/spatial-geometry/convexhullaggregate-geometry-data-type?view=sql-server-ver16) (I have no idea what convex hulls are but they sound very cool!).
  * NTS has [ConvexHull.Create](https://nettopologysuite.github.io/NetTopologySuite/api/NetTopologySuite.Algorithm.ConvexHull.html) which seems to correspond.
  * PostGIS and Spatialite don't have this, but the [PostGIS docs](https://postgis.net/docs/ST_ConvexHull.html) suggest doing simply `ST_ConvexHull(ST_Collect(geom))`. We can do the same for Spatialite.

**Comment by airbreather** [comment]
> * CollectionAggregate:
>     * For two points (0,0) and (1,1), NTS [GeometryCombiner.Combine](https://nettopologysuite.github.io/NetTopologySuite/api/NetTopologySuite.Geometries.Utilities.GeometryCombiner.html) returns `MULTIPOINT ((0 0), (1 1))` whereas SQL Server [CollectionAggregate](https://docs.microsoft.com/en-us/sql/t-sql/spatial-geometry/collectionaggregate-geometry-data-type?view=sql-server-ver16) returns `GEOMETRYCOLLECTION (POINT (0 0), POINT (1 1))`. This is probably OK since NTS MultiPoint extends GeometryCollection, and GeometryCombiner.Combine's return type is Geometry.
>     * A slightly more faithful translation for CollectionAggregate may be the GeometryCollection [constructor](https://nettopologysuite.github.io/NetTopologySuite/api/NetTopologySuite.Geometries.GeometryCollection.html#constructors), but unfortunately that accepts an array, not an IEnumerable (so we'd need to recognize and ignore ToArray over that). Translating constructor calls is also less easy.

Since these aggregate functions don't look like they're standardized in [06-104r4](https://www.ogc.org/standards/sfs) like most other things tend to be*, not to mention how many decades these different tools have each been doing their own thing, I think some amount of unfaithfulness when trying to translate between these same NTS calls and all of these different providers.

*If there's another standard I don't know about that **everybody else but NTS** is following, I'm sure we can add something to our core library to help facilitate things like this.

As long as the `Geometry` that NTS would produce is [**topologically equal**](http://nettopologysuite.github.io/NetTopologySuite/api/NetTopologySuite.Geometries.Geometry.html#NetTopologySuite_Geometries_Geometry_EqualsTopologically_NetTopologySuite_Geometries_Geometry_) to the `Geometry` that the provider gives us, then I think that's acceptable (though admittedly some users will have headaches with it).

> * EnvelopeAggregate:
>   * Weirdly, NTS [EnvelopeCombiner.Combine](https://nettopologysuite.github.io/NetTopologySuite/api/NetTopologySuite.Geometries.Utilities.EnvelopeCombiner.html#NetTopologySuite_Geometries_Utilities_EnvelopeCombiner_Combine_System_Collections_Generic_IEnumerable_NetTopologySuite_Geometries_Geometry__) returns an [Envelope](https://nettopologysuite.github.io/NetTopologySuite/api/NetTopologySuite.Geometries.Envelope.html), which is not a Geometry. Note that [Geometry.Envelope](https://nettopologysuite.github.io/NetTopologySuite/api/NetTopologySuite.Geometries.Geometry.html#NetTopologySuite_Geometries_Geometry_Envelope) does return Geometry, which is why we can translate the non-aggregate version ([test](https://github.com/dotnet/efcore/blob/main/test/EFCore.Specification.Tests/Query/SpatialQueryTestBase.cs?rgh-link-date=2022-05-27T12%3A55%3A50Z#L548)).
>    * It's possible to convert an Envelope into a Geometry via [GeometryFactory.ToGeometry](https://nettopologysuite.github.io/NetTopologySuite/api/NetTopologySuite.Geometries.GeometryFactory.html#NetTopologySuite_Geometries_GeometryFactory_ToGeometry_NetTopologySuite_Geometries_Envelope_), but that makes the C#-side construct really complicated. We should probably wait for requests before looking into this.

Agreed: given that not all providers (no providers?) natively have the concept of "Envelope" a distinct entity, this seems fairly wacky to implement and not-very-discoverable from the perspective of a user.

If there's a real user appetite for this, I wouldn't mind adding something like an `EnvelopeCombiner.CombineAsGeometry` that does basically the same thing that you're talking about here, just wrapped behind a single static method that you could detect.

> * Aggregate convex hull:
>     * [...] (I have no idea what convex hulls are but they sound very cool!).

Short version is, it's the (unique) smallest [convex polygon](https://en.wikipedia.org/wiki/Convex_polygon) that covers all of the input geometries.  There are other ways to get smaller polygons ([this one's](https://github.com/mapbox/concaveman) my favorite), but those won't be convex (short version of "convex" is that you can pick any two points in the polygon, and the line segment between them is fully covered by the polygon).

> * NTS has [ConvexHull.Create](https://nettopologysuite.github.io/NetTopologySuite/api/NetTopologySuite.Algorithm.ConvexHull.html) which seems to correspond.

Correct.  Or, at least, that was the **intent** of NetTopologySuite/NetTopologySuite@d826972, anyway :-)

**Comment by roji** [comment]
> If there's a real user appetite for this, I wouldn't mind adding something like an EnvelopeCombiner.CombineAsGeometry that does basically the same thing that you're talking about here, just wrapped behind a single static method that you could detect.

That would definitely help us out - we'd be able to translate that directly to its SQL equivalent (and I can imagnie it may even be useful in general to NTS users!).

> Aggregate convex hull:

Thanks for the explanation, am not at all a spatial guy :) And yeah, ConvexHull.Create fits our needs perfectly.

**Comment by roji** [comment]
Issue for addingEnvelopeCombiner.CombineAsGeometry to NTS: https://github.com/NetTopologySuite/NetTopologySuite/issues/610
